### ISSUE TITLE
Discussion of Past Year Paper 

### ISSUE THREAD

**Issue by ZhangAnli** [nus-cs2103-AY2021S2/forum#357]
Hello! Here are my answers for this PYP. Please feel free to let me know what my mistakes are! 

![image](https://user-images.githubusercontent.com/65236091/115832297-5161e180-a445-11eb-9780-3ea4c1d76996.png)

<img width="835" alt="Screen Shot 2021-04-23 at 3 06 28 PM" src="https://user-images.githubusercontent.com/65236091/115832477-879f6100-a445-11eb-8742-d445348be629.png">

**Comment by tlylt** [comment]
Thanks for sharing!
- Inventory
  - should include a dependency arrow to Billable
  - I would use an arrow to indicate navigability from Inventory to Item
- StockItem
  - I would include the constructor
  - should include a dependency arrow to Rating
- Review
  - I would include the constructor
  - StockItem to Review, I used 0..1 instead as I feel that it might be possible if null is passed in, might need prof to confirm this
  - same as above for association to rating, I used 0..1
- Billable
  - If I'm not wrong, all methods are public within interface

PS: Just my 2 cents :)

**Comment by ong6** [comment]
To add to what @tlylt has mentioned,

I think it's possible for StockItem to have a composition to Review (since it seems unlikely that a review would exist without an item)

My attempt:
![image](https://user-images.githubusercontent.com/19247856/115840522-a1eb2780-a3d0-11eb-9738-f326ff8d7be8.png)

**Comment by tlylt** [comment]
> To add to what @tlylt has mentioned,
> 
> I think it's possible for StockItem to have a composition to Review (since it seems unlikely that a review would exist without an item)
> 
> My attempt:
> ![image](https://user-images.githubusercontent.com/19247856/115840522-a1eb2780-a3d0-11eb-9738-f326ff8d7be8.png)

Btw your double arrow from Inventory to Item is wrong. If there is a stronger association indicated by the solid line, there is no need to include a dependency line anymore

**Comment by kouyk** [comment]
> StockItem to Review, I used 0..1 instead as I feel that it might be possible if null is passed in

Even if null is passed in, a Review object will still be created, so it should be just 1.

**Comment by ong6** [comment]
> Btw your double arrow from Inventory to Item is wrong. If there is a stronger association indicated by the solid line, there is no need to include a dependency line anymore

Oops forgot to erase it

**Comment by tlylt** [comment]
> > StockItem to Review, I used 0..1 instead as I feel that it might be possible if null is passed in
> 
> Even if null is passed in, a Review object will still be created, so it should be just 1.

@kouyk True, I think in this case StockItem to Review has to be 1. I think for Review to Rating, it could still be 0..1?Unless it does not matter if it is null or not

**Comment by ong6** [comment]
> > > StockItem to Review, I used 0..1 instead as I feel that it might be possible if null is passed in
> > 
> > 
> > Even if null is passed in, a Review object will still be created, so it should be just 1.
> 
> @kouyk True, I think in this case StockItem to Review has to be 1. I think for Review to Rating, it could still be 0..1

Review to rating should also be 1, since the code indicates that there is a Rating inside review. I don't think null matters in this case. Maybe @damithc  can clarify more?

**Comment by tlylt** [comment]
![image](https://user-images.githubusercontent.com/41845017/115842136-1e711b00-a450-11eb-9a4a-9ff02daa9796.png)
If Review class is simply like above, we all agree that it could be 0..1,
So either it means that if a variable is never explicitly initialized we can treat it as 0..1
or it means that if a variable could be assigned null, then we treat it as 0..1
That's my interpretation, will be good if prof clarifies

**Comment by damithc** [comment]
> Review to rating should also be 1, since the code indicates that there is a Rating inside review. I don't think null matters in this case. Maybe @damithc can clarify more?

Wasn't this entire question discussed during tutorial 8? You may want to watch the week 8 tutorial recording. :-p

**Comment by tlylt** [comment]
> > Review to rating should also be 1, since the code indicates that there is a Rating inside review. I don't think null matters in this case. Maybe @damithc can clarify more?
> 
> Wasn't this entire question discussed during tutorial 8? You may want to watch the week 8 tutorial recording. :-p

Are tutorial recordings available for students to view? Doesn't seem to be able to find them.

**Comment by damithc** [comment]
> Are tutorial recordings available for students to view? Doesn't seem to be able to find them.

In the MS Team for your tutorial group, there is a file called `Recordings.docx` that should contain all the tutorial recordings. If not found, ask the tutor to update it.

As the exam is tomorrow, here's the recording from my T8.

https://nus-sg.zoom.us/rec/share/Nk2-lh73_3zm5ij8kf5cipZwc1zpdWwrIVNz8L5Wki_I-Fq7MpV5P1KIjH4oqZ_H.h-TLRgoVjgwFsOSV 

Access Passcode: cs2103-tutorial

**Comment by ong6** [comment]
According to the tut, 0..1 is fine for variables (40:00 mins) prof mentions this. 

![image](https://user-images.githubusercontent.com/19247856/115858208-850c1f80-a3e3-11eb-9a64-452a8fca3671.png)

**Comment by prerthan99** [comment]
here's my attempt.  ive fixed some of the 'bugs' but i think its pretty similar

![Screenshot 2021-04-23 at 5 21 32 PM](https://user-images.githubusercontent.com/54735177/115876524-019d0d80-a479-11eb-9379-a071bf20f5d4.jpg)

**Comment by prerthan99** [comment]
actually, I do think that from StockItem to Rating there should be 1 because it's in the constructor.  Im not sure if we can pass in null instead of the enums in a constructor without causing an error.  if it is possible to pass null then I guess 0..1 is valid but if we cant pass in null, id say it should be 1

**Comment by prerthan99** [comment]
> > Are tutorial recordings available for students to view? Doesn't seem to be able to find them.
> 
> In the MS Team for your tutorial group, there is a file called `Recordings.docx` that should contain all the tutorial recordings. If not found, ask the tutor to update it.
> 
> As the exam is tomorrow, here's the recording from my T8.
> 
> https://nus-sg.zoom.us/rec/share/Nk2-lh73_3zm5ij8kf5cipZwc1zpdWwrIVNz8L5Wki_I-Fq7MpV5P1KIjH4oqZ_H.h-TLRgoVjgwFsOSV
> 
> Access Passcode: cs2103-tutorial

is the audio working or is it just my issue?

**Comment by kouyk** [comment]
> actually, I do think that from StockItem to Rating there should be 1 because it's in the constructor. 

Isn't StockItem to Rating a dependency? Where does the multiplicity come in?

> Im not sure if we can pass in null instead of the enums in a constructor without causing an error.

Enums can be set to null, it is the responsibility of the relevant code to verify that it is not null to ensure correctness.

**Comment by kouyk** [comment]
> is the audio working or is it just my issue?

It does appear that the issue is on your side, the audio works for me.
